corehq/project_limits/rate_limiter.py

```diff
@@ -28,62 +29,63 @@ class RateLimiter(object):
     ...     my_feature_rate_limiter.report_usage('my_domain')
 
     """
-    def __init__(self, feature_key, get_rate_limits, scope_length=1):
+    def __init__(self, feature_key, get_rate_limits):
         self.feature_key = feature_key
         self.get_rate_limits = get_rate_limits
-        self.scope_length = scope_length
-
-    def get_normalized_scope(self, scope):
-        if scope is None:
-            scope = ()
-        elif isinstance(scope, str):
-            scope = (scope,)
-        elif not isinstance(scope, tuple):
-            raise ValueError("scope must be a string or tuple: {!r}".format(scope))
-        elif len(scope) != self.scope_length:
-            raise ValueError("The scope for this rate limiter must be of length {!r}"
-                             .format(self.scope_length))
-        return scope
-
-    def report_usage(self, scope=None, delta=1):
-        scope = self.get_normalized_scope(scope)
-        for rate_counter, limit in self.get_rate_limits(*scope):
-            rate_counter.increment((self.feature_key,) + scope, delta=delta)
-
-    def get_window_of_first_exceeded_limit(self, scope=None):
-        for rate_counter_key, current_rate, limit in self.iter_rates(scope):
-            if current_rate >= limit:
-                return rate_counter_key
+
+    def report_usage(self, scope='', delta=1):
+        for limit_scope, limits in self.get_rate_limits(scope):
+            for rate_counter, limit in limits:
+                rate_counter.increment(self.feature_key + limit_scope, delta=delta)
+
+    def get_window_of_first_exceeded_limit(self, scope=''):
+        for limit_scope, rates in self.iter_rates(scope):
+            for rate_counter_key, current_rate, limit in rates:
+                if current_rate >= limit:
+                    return rate_counter_key
 
         return None
 
-    def allow_usage(self, scope=None):
-        return all(current_rate < limit
-                   for rate_counter_key, current_rate, limit in self.iter_rates(scope))
+    def allow_usage(self, scope=''):
+        allowed = False
+        for limit_scope, rates in self.iter_rates(scope):
+            allow = all(current_rate < limit
+                        for rate_counter_key, current_rate, limit in rates)
+            # allow usage if any limiter is below the threshold
```

Comment:
@calellowitz I think this should read 'all' not 'any'

Reply:
I think this is probably a confusing comment, but "any" is what I intended. It's all of the inner loop (rates for different time frames within a single limiter), but any of the outer loop (limit set). So if they are under the domain limit but over the enterprise limit, it is still allowed, but you are considered over the domain limiter if you are over the hour limit without being over the minute limit.

Reply:
https://github.com/dimagi/commcare-hq/pull/32025